src/util/createInstantSearchComponent.js

```diff
@@ -34,6 +34,16 @@ export const createInstantSearchComponent = component =>
           // private InstantSearch.js API:
           this.instantSearchInstance._searchFunction = searchFunction;
         },
+        middlewares: {
+          immediate: true,
+          handler(next, prev) {
+            (next || [])
```

Comment:
the previous middleware that no longer are in next middleware (opposite filter) should call a method on InstantSearch that removes the middleware reference from the list and calls its dispose, very similar to removeWidgets